system/mp/mpi_wrapper.cpp

```
@@ -0,0 +1,513 @@
+//            if (completed || error) //unlikely an error would occur
+//            {
+//                delete _popCommData(i);
+//            }
+//        }
+//    }
+
+    asyncCommData.push_back(commData);
+}
+
+CommData* popCommData(int index)
+{
+    _TF("popCommData", index);
+    CriticalBlock block(commDataLock);
+    return _popCommData(index);
+}
+
+std::vector<CommData*> popCommData(bool send, int rank, int tag, MPI::Comm& comm)
+{
+   _TF("popCommData", rank, tag);
+   CriticalBlock block(commDataLock);
+   std::vector<CommData*> matchedResults = std::vector<CommData*>();
+   int size = asyncCommData.size();
+   for(int i=(size-1); i>=0; i--)
+   {
+       if (asyncCommData[i]->isEqual(send, rank, tag, comm))
+           matchedResults.push_back(_popCommData(i));
+   }
+   return matchedResults;
+}
+
+CommData* popCommData(CommData * commData)
+{
+    _TF("popCommData(CommData * commData)", commData->rank, commData->tag);
+    CriticalBlock block(commDataLock);
+    int size = asyncCommData.size();
+    for(int i=(size-1); i>=0; i--)
+    {
+        if (asyncCommData[i]==commData)
+        {
+            _popCommData(i);
+            break;
+        }
+    }
+    return commData;
+}
+
+//----------------------------------------------------------------------------------------------------------------------------//
+
+#define TAG_UB 100000 //INT_MAX //MPI_TAG_UB
+int getRank(rank_t sourceRank)
+{
+    _TF("getRank", sourceRank);
+    if (sourceRank == RANK_ALL)
+        return MPI::ANY_SOURCE;
+    else
+        return sourceRank;
+}
+
+int getTag(mptag_t mptag)
+{
+    _TF("getTag", mptag);
+    int returnVal;
+    unsigned tag = mptag;
+    if (mptag == TAG_ALL)
+        returnVal = MPI::ANY_TAG;
+    else
+    {
+        if (tag > TAG_UB)
+        {
+            tag= TAG_UB-((unsigned)-tag);                         //MPI doesn't allow custom tags with negative values or beyond MPI_TAG_UB. Thus we shift it to the MPI_TAG_UB range
```

This isn't always going to work..

If will work if mptag is a negative int between -1 and -TAG_UB,
but it will not work if mptag is >=TAG_UB <= ((unsigned)-TAG_UB)

e.g. if TAG_UB is 100000 and mptag is 100123

then the result will be = 200123 , i.e. bigger than TAG_UB

---

Now it may be that we can reasonably assert that no mptag is bigger than TAG_UB, or at least for now.
We could use 1000's but hopefully never use a 100,000 tags.
But need to careful about the negative replytag wrapping too.

And I can't guarantee that nothing in the system uses mptag_t's now that are > TAG_UB.
If they do, we'll have to revisit the issue.

But the above should check and fail/assert if anything is out of range, rather than continue.

---

I think I'd code it like this:
```
#define SPECIAL_TAG_BASE (TAG_UB/2) // reserve half range for special tags (inc. reply tags)
int getMPITag(mptag_t mptag)
{
    _TF("getTag", mptag);

    if (mptag == TAG_ALL)
        return MPI::ANY_TAG;
    else
    {
        int tag = (int)mptag;
        if (tag < 0)
        {
            assertex(-tag < TAG_UB-SPECIAL_TAG_BASE); // check not out of range
            return -tag + SPECIAL_TAG_BASE;
        }
        else
        {
            assertex(tag <= TAG_UB);
            return tag;
        }
    }
}
```